doc/developer/code-coverage.md

```diff
@@ -0,0 +1,50 @@
+# Code Coverage for Pull Requests
+
+## Motivation
+
+A full [code coverage report](http://65.109.125.29/coverage_01879007-43c4-49db-a81d-a44e93dfb55a/) allows you to check for each part of the code if there is a test that currently covers it by executing the relevant lines of code. Unfortunately the motivation to increase code coverage using full reports is low, and the effort to understand why no existing test covers the code in question is high.
+
+Thus to prevent regressions and new bugs we instead want to ensure that new and changed productive code in a pull request (PR) is adequately tested. This makes sense both when code is refactored as well as when new functionality is added. For each PR you or someone from the QA team can use the [Coverage](https://buildkite.com/materialize/coverage/builds) Buildkite pipeline to trigger coverage for it.
+
+## Triggering Coverage for your PR
+
+As an example let's take the *[adapter: Update privileges when an owner is updated](https://github.com/MaterializeInc/materialize/pull/18777)* pull request. To run PR code coverage against it, go to the [Coverage Buildkite pipeline](https://buildkite.com/materialize/coverage), click on "New Build" and enter the full commit hash and branch with your Github username:
+
+![Commit: 69de84971e8c16ab459c88be73870bb676dbc478, Branch: jkosh44:update-privilege-owners](assets/coverage-new-build.png)
+
+After some time your [Coverage run](https://buildkite.com/materialize/coverage/builds/26) will be finished and contain a summary of any uncovered lines in the form of a `git diff` where each removed line is not covered:
+
+![Uncovered Lines in Pull Request](assets/coverage-uncovered-lines.png)
+
+We can see that some of the newly added code is not covered. This is the point where we should think about if we can write an explicit test case that tests this code adequately.
+
+Additionally you can also download the full coverage report in HTML format from the artifacts of the "Analyze code coverage for PR" step. This should only be necessary if you are interested in seeing coverage of the entire product:
+
+![Artifacts](assets/coverage-artifacts.png)
+
+## Limitations
+
+The [Coverage](https://buildkite.com/materialize/coverage/builds) Buildkite pipeline runs a subset of tests from the [Tests](https://buildkite.com/materialize/tests) pipeline, which is already automatically run against each pull request. The excluded tests are marked with a `coverage: skip` in the [`pipeline.template.yml`](https://github.com/MaterializeInc/materialize/blob/main/ci/test/pipeline.template.yml):
+
+- Cargo test: Code should be tested in a real environment using SQL/Kafka/etc., unit test can give a false sense of seceurity
```

Comment:
I don't fully agree with "unit test can give a false sense of security" and would rephrase it to something like that we also want to have the code seen tested in interaction.